src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpLogSerializer.cs

```diff
@@ -47,13 +47,36 @@ internal static int WriteLogsData(byte[] buffer, int writePosition, SdkLimitOpti
             logRecords.Add(logRecord);
         }
 
-        writePosition = WriteResourceLogs(buffer, writePosition, sdkLimitOptions, experimentalOptions, resource, ScopeLogsList);
+        writePosition = TryWriteResourceLogs(buffer, writePosition, sdkLimitOptions, experimentalOptions, resource, ScopeLogsList);
         ProtobufSerializer.WriteReservedLength(buffer, logsDataLengthPosition, writePosition - (logsDataLengthPosition + ReserveSizeForLength));
         ReturnLogRecordListToPool();
 
         return writePosition;
     }
 
+    internal static int TryWriteResourceLogs(byte[] buffer, int writePosition, SdkLimitOptions sdkLimitOptions, ExperimentalOptions experimentalOptions, Resources.Resource? resource, Dictionary<string, List<LogRecord>> scopeLogs)
+    {
+        try
+        {
+            writePosition = WriteResourceLogs(buffer, writePosition, sdkLimitOptions, experimentalOptions, resource, scopeLogs);
+        }
+        catch (IndexOutOfRangeException)
+        {
+            if (!ProtobufSerializer.IncreaseBufferSize(ref buffer, OtlpSignalType.Logs))
+            {
+                throw;
+            }
+
+            return TryWriteResourceLogs(buffer, writePosition, sdkLimitOptions, experimentalOptions, resource, scopeLogs);
```

Comment:
should we add any protections against an infinite loop here?
ie: after 5 retries, quit?

Reply:
We have that logic as part of `IncreaseBufferSize`. After seven attempts, it will return false, and we throw an exception.

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/44bc90a25b9f6d5323bebc6883ec034ccc36c721/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufSerializer.cs#L344
I don't see any code tracking "seven attempts", but this method does have some safeguards for a MaxBufferSize as well as a `catch (OutOfMemoryException)`. Either of these would break out of any infinite loop so I think we're safe here.

Reply:
More than infinite loop it will cause StackOverflowException and crash the process. I added a detailed comment with Trace implementation and felt not to duplicate across all the signals. Ref: https://github.com/open-telemetry/opentelemetry-dotnet/blob/44bc90a25b9f6d5323bebc6883ec034ccc36c721/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpTraceSerializer.cs#L60